index.html

```diff
@@ -292,7 +292,38 @@ <h3>
       <li>Let |providers| be |options|'s {{CredentialRequestOptions/digital}}'s
       {{DigitalCredentialRequestOptions/providers}} member.
       </li>
-      <li>If |providers| is empty, [=exception/throw=] a {{TypeError}}.
+      <li>If |providers| [=list/is empty=], [=exception/throw=] a
+      {{TypeError}}.
+      </li>
+      <li>Let |range| be [=list/get the indices|the indices=] of |providers|.
+      </li>
+      <li>[=Set/For each=] |n| of the |range|:
+        <ol>
+          <li>Let |provider| be the [=list/item=] at |providers|'s |n| index.
+          </li>
+          <li>If |provider|'s {{DigitalCredentialsProvider/protocol}} does not
+          match a [=digital credentials registry/protocol identifier=] from the
+          [=digital credentials registry=], [=exception/throw=] a
+          {{TypeError}}.
+          </li>
+          <li>Let |dictionary| be the [=digital credentials registry/Web IDL
+          dictionary=] for |provider|'s
+          {{DigitalCredentialsProvider/protocol}}.
+          </li>
+          <li>Let |request| be |provider|'s
+          {{DigitalCredentialsProvider/request}} [=converted to an idl
+          value|converted=] to |dictionary|. [=exception/throw|Rethrow=] any
+          exceptions.
+          </li>
+          <li>Validate |request| against any validation rules set forth in the
+          corresponding [=digital credentials registry/specification=]. If any
+          of |request|'s members are invalid, [=exception/throw=] a
```

Comment:
Right, but we can validate a lot of things when the request is passed to the browser. I agree that some aspects cannot be validated, and that can be left up to the wallet upon decrypting the request (if the request, or parts of it, is encrypted in the first place). 

At the same time, we need the assurances of what to do when the user agents receive a request in a particular format.

Reply:
> At the same time, we need the assurances of what to do when the user agents receive a request in a particular format.

My intuition is that there is a subset of the request that the browser cares about and should validate which is the query language, specifically, the [`presentation_definition`](https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#name-presentation_definition-par) parameter in OpenID4VP which takes shape as a [Presentation Exchange](https://identity.foundation/presentation-exchange/).

That seems like something that the browser uses to help the user select which wallet has the credential that's being requested, as well as the privacy and security consequences of sharing that credential.

Beyond that, I think it would be harmful (from a security standpoint) to validate (e.g. drop an unknown parameter) the rest of the parameters. 

Looking at the list of parameters in the [Authorization Request](https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#name-authorization-request) we have the following, in addition to `presentation_definition`:

- `client_id_scheme`
- `client_metadata`
- `nonce`
- `scope`
- `response_mode`

None of these seem like things that the browser needs to form an opinion on, right? 

> if the request, or parts of it, is encrypted in the first place

As soon as we introduce any parameter that is a `DOMString` (for example, `nonce`), verifiers and wallets can communicate freely through that channel. So, there is a escape hatch for any validation that holds the protocol designers and deployment back.

From a security standpoint, I think that's a feature, not a bug: if OpenID4VP (or developers running servers) figure out a safer way to do things, they should be able to develop that without having to ask for browsers to catch up. For example, say there is a production fire happening somewhere, you we wouldn't our browser users to wait until all browsers are patched and redeployed to fix the issue, right?

Reply:
I share much of @samuelgoto's intuition FWIW. But also just want to note that there's work underway in OpenID4VP towards developing a new query language alternative to PE https://github.com/openid/OpenID4VP/pull/220

Reply:
I wrote down a proposal here: https://github.com/WICG/digital-credentials/issues/161#issue-2481939110

Reply:
> As soon as we introduce any parameter that is a DOMString (for example, nonce), verifiers and wallets can communicate freely through that channel.

I don’t think that’s true. We could certainly enforce restrictions or other checks if we see ab/misuse. 

Playing whack-a-mole is not fun, but it’s absolutely something we would do and have done in the past. 

> Beyond that, I think it would be harmful (from a security standpoint) to validate (e.g. drop an unknown parameter) the rest of the parameters.

Potentially, yes. If the request was signed along with any non-standard parameters (which would get dropped by the browser before being passed to the wallet). At the same time, non-standard parameters shouldn’t be part of the request. An RP will likely need some means to check if the browser supports the request structure. 

My intuition, and because of our security architecture, WebKit will drop anything non-standard.

Reply:
Spoke @samuelgoto about how this would be implemented in practice, and agreeing with @RByers that this should happen in the wallet or as it's being passed to the wallet, and not for all the things that are passed in.

Reply:
I'll try to update this PR soon!